### PR TITLE
Improve concurrency for needed parts.

### PR DESCRIPTION
**Pull request by thinker0**
Improve concurrency for needed parts.
- Change concurrentHashMap
- Improve thread safe OutputCollector
- Improve thread safe IOutputCollector

#3104

### CODE REVIEW

**Comment by nwangtw** [comment]
That is the root cause of the exception? Nice find!

**Comment by thinker0** [comment]
@nwangtw 

I created a Bolt with MoultiThread (finagle-http) async and ack all at once.

I think NPE is happening often.

heron-0.18.7 I can not create a maven package file, so I can not test it in my project.
Can you tell me how to package it so I can test it?

**Comment by nwangtw** [comment]
I see. Make sense then when you use multi threading. In this case you might be better to synchronized {} to make sure ack() is not called many times at the same time. OutputCollector has a queue in it and I am not sure the queue is thread-safe. In Heron engine, the core assumes single thread environment.

Internally, we build library and use it directly, instead of the ones in Maven.

bazel build --compilation_mode=dbg --config=darwin scripts/packages:tarpkgs

File: heron/api/src/java/org/apache/heron/api/metric/MultiCountMetric.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ public CountMetric scope(String key) {
 
   @Override
   public Map<String, Long> getValueAndReset() {
-    Map<String, Long> ret = new HashMap<>();
+    Map<String, Long> ret = new ConcurrentHashMap<>();
```
Comment (nwangtw) [inline heron/api/src/java/org/apache/heron/api/metric/MultiCountMetric.java]: ConcurrentHashMap might not be necessary in getValueAndReset(). This function is called every X seconds to sample the current metrics and export by Heron instance, so it is not called from multiple threads.

File: heron/api/src/java/org/apache/heron/api/metric/CountStatAndMetric.java [hunk 2]
```diff
@@ -176,7 +176,7 @@ private synchronized void rotate(long value, long timeSpent,
   }
 
   synchronized Map<String, Long> getTimeCounts(long now) {
-    Map<String, Long> ret = new HashMap<>();
+    Map<String, Long> ret = new ConcurrentHashMap<>();
```
Comment (nwangtw) [inline heron/api/src/java/org/apache/heron/api/metric/CountStatAndMetric.java]: ConcurrentHashMap  is not needed here I think. This is a get function and it is already synchronized.

File: heron/api/src/java/org/apache/heron/api/metric/LatencyStatAndMetric.java [hunk 3]
```diff
@@ -225,7 +225,7 @@ private synchronized void rotate(long lat, long count, long timeSpent, long targ
   }
 
   synchronized Map<String, Double> getTimeLatAvg(long now) {
-    Map<String, Double> ret = new HashMap<>();
+    Map<String, Double> ret = new ConcurrentHashMap<>();
```
Comment (nwangtw) [inline heron/api/src/java/org/apache/heron/api/metric/LatencyStatAndMetric.java]: ConcurrentHashMap  is not needed here I think. This is a get function and it is already synchronized.

File: storm-compatibility/src/java/org/apache/storm/metric/api/MultiCountMetric.java [hunk 4]
```diff
@@ -36,7 +36,7 @@ public CountMetric scope(String key) {
   }
 
   public Object getValueAndReset() {
-    Map<String, Object> ret = new HashMap<>();
+    Map<String, Object> ret = new ConcurrentHashMap<>();
```
Comment (nwangtw) [inline storm-compatibility/src/java/org/apache/storm/metric/api/MultiCountMetric.java]: HashMap should be good here I think.

File: storm-compatibility/src/java/org/apache/storm/metric/api/MultiReducedMetric.java [hunk 5]
```diff
@@ -39,7 +39,7 @@ public ReducedMetric scope(String key) {
   }
 
   public Object getValueAndReset() {
-    Map<String, Object> ret = new HashMap<>();
+    Map<String, Object> ret = new ConcurrentHashMap<>();
```
Comment (nwangtw) [inline storm-compatibility/src/java/org/apache/storm/metric/api/MultiReducedMetric.java]: HashMap should be good here I think.

**Comment by thinker0** [comment]
Please review the PR. @nwangtw

**Comment by joshfischer1108** [comment]
The changes after Ning's comments look fine to me.  I'll leave it to @nwangtw to approve.

**Comment by thinker0** [comment]
> I see. Make sense then when you use multi threading. In this case you might be better to synchronized {} to make sure ack() is not called many times at the same time. OutputCollector has a queue in it and I am not sure the queue is thread-safe. In Heron engine, the core assumes single thread environment.
> 
> Internally, we build library and use it directly, instead of the ones in Maven.
> 
> bazel build --compilation_mode=dbg --config=darwin scripts/packages:tarpkgs

Made the OutputCollector thread-safe.

@nwangtw,
```
synchronized
```
works fine without it.
And it works well in Production.

**Comment by thinker0** [comment]
@joshfischer1108 PR Check ^^

**Pull request merged**